### PR TITLE
Add general guidelines for Spark on Graviton

### PR DESCRIPTION
**Pull request by nafeabd**
*Issue #, if available:*

*Description of changes:*
Adding some general guidelines/recommendations that will be helpful for users trying to improve application performance across Spark clusters. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: DataAnalytics.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Spark on Graviton
+Below are some general guidelines that can be referenced by users trying to improve over all Application performance across Spark clusters. Note that, there is no magic/standard values that works in all use cases and each case can be different. Most of the times, it depends on dataset sizes, cores available, workload type etc.
```
Comment (AGSaidi) [inline DataAnalytics.md:2]: More introduction please about why this is difficult and what the goal should be.

The goal is generally to keep all the cpus busy and not have a small set of cores executing long running tasks. This depends on the datset size, number of instances, core count/instance, and computational complexity.

Should also have a bit of text on bottlenecks in Spark being cpu, vs disk vs. network

File: DataAnalytics.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Spark on Graviton
+Below are some general guidelines that can be referenced by users trying to improve over all Application performance across Spark clusters. Note that, there is no magic/standard values that works in all use cases and each case can be different. Most of the times, it depends on dataset sizes, cores available, workload type etc. 
+
+1. Shuffle Partitions: This config option is important to achieve optimal performance. Recommendations in regards to this are
```
Comment (AGSaidi) [inline DataAnalytics.md:4]: recommendations here, next line "it's recommended"

File: DataAnalytics.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    * It's recommended to have the partition size less than 200 MB to gain optimized performance
+    * Usually number of partitions should be multiples of number of cores available (1xcores, 2xcores .. etc)
+2. AWS EMR team recommends using EMR defaults. For any other specific cases that needs specific tunings, the general optimization guide is https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-performance.html.
+3. Spark official tuning guide - https://spark.apache.org/docs/latest/sql-performance-tuning.html. Also worth looking at Adaptive query execution (AQE) which was not enabled by default before Spark 3.2 (https://spark.apache.org/docs/latest/sql-performance-tuning.html#adaptive-query-execution).
```
Comment (AGSaidi) [inline DataAnalytics.md:8]: what does "also worth looking" mean? How is that actionalable?

File: DataAnalytics.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    * It's recommended to have the partition size less than 200 MB to gain optimized performance
+    * Usually number of partitions should be multiples of number of cores available (1xcores, 2xcores .. etc)
+2. AWS EMR team recommends using EMR defaults. For any other specific cases that needs specific tunings, the general optimization guide is https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-performance.html.
+3. Spark official tuning guide - https://spark.apache.org/docs/latest/sql-performance-tuning.html. Also worth looking at Adaptive query execution (AQE) which was not enabled by default before Spark 3.2 (https://spark.apache.org/docs/latest/sql-performance-tuning.html#adaptive-query-execution). 
+4. Our current results show Graviton3 instances to be more performant than Graviton2 instances. For users dealing with issues across performance on Graviton 2 instances, first recommendation is to try above tuning options and check the performance. In case of no improvement, we recommend to migrate to Graviton3 instances if possible. 
+5. Spark 3 is faster than Spark 2 in general. We recommend to use latest version of Spark along with latest Java version.
```
Comment (AGSaidi) [inline DataAnalytics.md:10]: again, very open ended and not particularly actionable. We've seen xx% delta between spark x and y? we recommend you use java 17? what?

File: DataAnalytics.md (hunk 1 above)
Comment (RamaMalladiAWS) [inline DataAnalytics.md:2]: Can you add perf data and also the version of Spark used in your tests/ runs?

File: README.md [hunk 5]
```diff
@@ -22,6 +22,7 @@ This repository provides technical guidance for users and developers using [Amaz
 * [Finding and managing AMIs for Graviton, with AWS SystemManager or CloudFormation](amis_cf_sm.md)
 * [DPDK, SPDK, and other datapath software](dpdk_spdk.md)
 * [TensorFlow](machinelearning/tensorflow.md)
+* [Spark on Graviton](DataAnalytics.md)
```
[the remark this answers is not in the document]
Comment (nafeabd) [reply README.md:25]: Idea is to have one section for different data processing frameworks like Spark. Currently it's only Spark, but we can add more in the future.

File: DataAnalytics.md (hunk 1 above)
Comment (nafeabd) [reply DataAnalytics.md]: Added more details in next revision

File: DataAnalytics.md (hunk 2 above)
Comment (nafeabd) [reply DataAnalytics.md]: Added perf data showing the delta in next revision

File: DataAnalytics.md (hunk 3 above)
Comment (nafeabd) [reply DataAnalytics.md]: Modified the text. Hope it is clear now.

File: DataAnalytics.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+1. Shuffle Partitions: This config option is important to achieve optimal performance. Recommendations in regards to this are
+    * It's recommended to have the partition size less than 200 MB to gain optimized performance
+    * Usually number of partitions should be multiples of number of cores available (1xcores, 2xcores .. etc)
+2. AWS EMR team recommends using EMR defaults. For any other specific cases that needs specific tunings, the general optimization guide is https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-performance.html.
+3. Spark official tuning guide - https://spark.apache.org/docs/latest/sql-performance-tuning.html. Also worth looking at Adaptive query execution (AQE) which was not enabled by default before Spark 3.2 (https://spark.apache.org/docs/latest/sql-performance-tuning.html#adaptive-query-execution). 
+4. Our current results show Graviton3 instances to be more performant than Graviton2 instances. For users dealing with issues across performance on Graviton 2 instances, first recommendation is to try above tuning options and check the performance. In case of no improvement, we recommend to migrate to Graviton3 instances if possible.
```
[the remark this answers is not in the document]
Comment (nafeabd) [reply DataAnalytics.md]: Removed in latest revision

File: DataAnalytics.md (hunk 4 above)
Comment (nafeabd) [reply DataAnalytics.md]: Added additional details in next revision.

File: DataAnalytics.md (hunk 1 above)
Comment (nafeabd) [reply DataAnalytics.md]: I added extra details to show the versions used. I avoided adding too many details w.r.t to config options though

File: DataAnalytics.md [hunk 7]
```diff
@@ -1,11 +1,34 @@
...
-5. Spark 3 is faster than Spark 2 in general. We recommend to use latest version of Spark along with latest Java version.
+Apache Spark is a data processing framework widely used to extract information from large pools of data.
+One main problem that affects performance is the straggler, where a long-running task slows down the entire cluster. Stragglers in Spark are usually caused by non-uniform distribution of work or data being skewed non uniformly across nodes, resulting in a single task taking up more work. Our goal should be to keep all the CPUs busy and not have a small set of cores executing long running tasks. Setting up a correct configuration depends on the dataset size, number of instances, core count/instance, and computational complexity.
 
+Below are some general guidelines that can be referenced by users trying to improve over all Application performance across Spark clusters. Since there is no one standard config set that works well in all cases, we advise users to benchmark real applications after following the below guidelines.
```
Comment (AGSaidi) [inline DataAnalytics.md:6]: "over all" -> overall; lower case application

File: DataAnalytics.md [hunk 8]
```diff
@@ -1,11 +1,34 @@
...
+|256	|39	|-78%	|
+|512	|37	|-79%	|
+|1024	|35	|-80%	|
+|2000	|35	|-80%	|
+
+*Lower numbers are better, and negative % diff means faster. Benchmarked on Spark 3.3.0 with Java 17 using spark-sql-perf framework from Databricks*
+
+
+
+2. When using Amazon EMR to setup Spark cluster, it is recommended to use EMR defaults for configuration options. For any other specific cases that need specific tuning, the general optimization guide can be referenced from https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark-performance.html.
```
Comment (AGSaidi) [inline DataAnalytics.md:32]: doesn't look like this 2. formats right compared to the 1. above?

File: DataAnalytics.md [hunk 9]
```diff
@@ -1,11 +1,34 @@
...
+3. Adaptive Query Execution(AQE) is an optimization technique in Spark SQL that makes use of runtime statistics to choose the most efficient query execution plan, which is enabled by default since Apache Spark 3.2.0. For users, using older Spark versions, we recommend checking the performance by turning it on (https://spark.apache.org/docs/latest/sql-performance-tuning.html#adaptive-query-execution)
```
Comment (AGSaidi) [inline DataAnalytics.md:33]: we recommend turning it on and seeing if it improves performance?

File: DataAnalytics.md [hunk 10]
```diff
@@ -1,11 +1,34 @@
...
+4. We have seen 40% improvement in performance when using Spark 3 with Java 17 when compared to Spark 2 with Java 8. So we recommend to use latest Spark 3 with Java 17.
```
Comment (AGSaidi) [inline DataAnalytics.md:34]: recommend to use -> recommend using

File: DataAnalytics.md (hunk 8 above)
Comment (nafeabd) [reply DataAnalytics.md:32]: I slightly adjusted content above. 1 and 2 have similar alignment. Only table is left aligned (github markdown defaults to left alignment for tables and doesn't allow to center align the table)

**Pull request merged**